snap/internal/file.go

```diff
@@ -0,0 +1,42 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package internal
+
+import (
+	"os"
+)
+
+// SizedFile holds an os.File plus its (initial) size.
+type SizedFile struct {
+	*os.File
+	size int64
+}
+
+func NewSizedFile(f *os.File) (*SizedFile, error) {
```

Comment:
The fact that this is just the initial size makes me wonder if we should tweak it a little bit to get the actual size. Something like this maybe?
```
diff --git a/snap/container.go b/snap/container.go
index 4349fb375b..fab495f5d4 100644
--- a/snap/container.go
+++ b/snap/container.go
@@ -38,7 +38,7 @@ type Container interface {
        RandomAccessFile(relative string) (interface {
                io.ReaderAt
                io.Closer
-               Size() int64
+               Size() (int64, error)
        }, error)
 
        // ReadFile returns the content of a single file from the snap.
diff --git a/snap/internal/file.go b/snap/internal/file.go
index 9f05ad5380..79cb20717b 100644
--- a/snap/internal/file.go
+++ b/snap/internal/file.go
@@ -26,17 +26,16 @@ import (
 // SizedFile holds an os.File plus its (initial) size.
 type SizedFile struct {
        *os.File
-       size int64
 }
 
 func NewSizedFile(f *os.File) (*SizedFile, error) {
+       return &SizedFile{File: f}, nil
+}
+
+func (f *SizedFile) Size() (int64, error) {
        fi, err := f.Stat()
        if err != nil {
-               return nil, err
+               return 0, err
        }
-       return &SizedFile{File: f, size: fi.Size()}, nil
-}
-
-func (f *SizedFile) Size() int64 {
-       return f.size
+       return fi.Size(), nil
 }
diff --git a/snap/snapdir/snapdir.go b/snap/snapdir/snapdir.go
index 5e47be7329..e52279906a 100644
--- a/snap/snapdir/snapdir.go
+++ b/snap/snapdir/snapdir.go
@@ -75,7 +75,7 @@ func (s *SnapDir) Install(targetPath, mountDir string, opts *snap.InstallOptions
 func (s *SnapDir) RandomAccessFile(file string) (interface {
        io.ReaderAt
        io.Closer
-       Size() int64
+       Size() (int64, error)
 }, error) {
        f, err := os.Open(filepath.Join(s.path, file))
        if err != nil {
diff --git a/snap/snapdir/snapdir_test.go b/snap/snapdir/snapdir_test.go
index ca3270b318..ece6f05dbc 100644
--- a/snap/snapdir/snapdir_test.go
+++ b/snap/snapdir/snapdir_test.go
@@ -82,7 +82,9 @@ func (s *SnapdirTestSuite) TestRandomAccessFile(c *C) {
        c.Assert(err, IsNil)
        defer r.Close()
 
-       c.Assert(r.Size(), Equals, int64(5))
+       sz, err := r.Size()
+       c.Assert(err, IsNil)
+       c.Assert(sz, Equals, int64(5))
 
        b := make([]byte, 2)
        n, err := r.ReadAt(b, 2)
diff --git a/snap/squashfs/squashfs.go b/snap/squashfs/squashfs.go
index 6590787fd2..75df9cd713 100644
--- a/snap/squashfs/squashfs.go
+++ b/snap/squashfs/squashfs.go
@@ -244,7 +244,7 @@ func (s *Snap) withUnpackedFile(filePath string, f func(p string) error) error {
 func (s *Snap) RandomAccessFile(filePath string) (interface {
        io.ReaderAt
        io.Closer
-       Size() int64
+       Size() (int64, error)
 }, error) {
        var f *os.File
        err := s.withUnpackedFile(filePath, func(p string) (err error) {
diff --git a/snap/squashfs/squashfs_test.go b/snap/squashfs/squashfs_test.go
index f510a4466d..73c245d458 100644
--- a/snap/squashfs/squashfs_test.go
+++ b/snap/squashfs/squashfs_test.go
@@ -328,7 +328,9 @@ func (s *SquashfsTestSuite) TestRandomAccessFile(c *C) {
        c.Assert(err, IsNil)
        defer r.Close()
 
-       c.Assert(r.Size(), Equals, int64(9))
+       sz, err := r.Size()
+       c.Assert(err, IsNil)
+       c.Assert(sz, Equals, int64(9))
 
        b := make([]byte, 4)
        n, err := r.ReadAt(b, 4)
```
but can be a followup of course (if we want it). I'm inclined to merge as is given that for our current use-cases it makes no difference. I'm a bit worried it may bite us down the road but then - YAGNI (maybe?)
```

Reply:
the code was like that but undid it on purpose see the comment to the commit https://github.com/snapcore/snapd/pull/9227/commits/fb087b94699cd51acf7b3fe14585c3b24f37873e

Reply:
Sorry that I overlooked this